connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala

```diff
@@ -0,0 +1,380 @@
+    val valueField = protoType.getMessageType.getFields.get(1)
+    val keyWriter = newWriter(keyField, keyType, protoPath :+ "key", catalystPath :+ "key")
+    val valueWriter =
+      newWriter(valueField, valueType, protoPath :+ "value", catalystPath :+ "value")
+    (updater, ordinal, value) =>
+      if (value != null) {
+        val messageList = value.asInstanceOf[java.util.List[com.google.protobuf.Message]]
+        val valueArray = createArrayData(valueType, messageList.size())
+        val valueUpdater = new ArrayDataUpdater(valueArray)
+        val keyArray = createArrayData(keyType, messageList.size())
+        val keyUpdater = new ArrayDataUpdater(keyArray)
+        var i = 0
+        messageList.forEach { field =>
+          {
+            keyWriter(keyUpdater, i, field.getField(keyField))
+            if (field.getField(valueField) == null) {
+              if (!valueContainsNull) {
+                throw QueryCompilationErrors.nullableArrayOrMapElementError(protoPath)
+              } else {
+                valueUpdater.setNullAt(i)
+              }
+            } else {
+              valueWriter(valueUpdater, i, field.getField(valueField))
+            }
+          }
+          i += 1
+        }
+        updater.set(ordinal, new ArrayBasedMapData(keyArray, valueArray))
+      }
+  }
+
+  /**
+   * Creates a writer to write Protobuf values to Catalyst values at the given ordinal with the
+   * given updater.
+   */
+  private def newWriter(
+      protoType: FieldDescriptor,
+      catalystType: DataType,
+      protoPath: Seq[String],
+      catalystPath: Seq[String]): (CatalystDataUpdater, Int, Any) => Unit = {
+    val errorPrefix = s"Cannot convert Protobuf ${toFieldStr(protoPath)} to " +
+      s"SQL ${toFieldStr(catalystPath)} because "
+    val incompatibleMsg = errorPrefix +
+      s"schema is incompatible (protoType = ${protoType} ${protoType.toProto.getLabel} " +
+      s"${protoType.getJavaType} ${protoType.getType}, sqlType = ${catalystType.sql})"
+
+    (protoType.getJavaType, catalystType) match {
+
+      case (null, NullType) => (updater, ordinal, _) => updater.setNullAt(ordinal)
+
+      // TODO: we can avoid boxing if future version of Protobuf provide primitive accessors.
+      case (BOOLEAN, BooleanType) =>
+        (updater, ordinal, value) => updater.setBoolean(ordinal, value.asInstanceOf[Boolean])
+
+      case (BOOLEAN, ArrayType(BooleanType, containsNull)) =>
+        newArrayWriter(protoType, protoPath, catalystPath, BooleanType, containsNull)
+
+      case (INT, IntegerType) =>
+        (updater, ordinal, value) => updater.setInt(ordinal, value.asInstanceOf[Int])
+
+      case (INT, ByteType) =>
+        (updater, ordinal, value) => updater.setByte(ordinal, value.asInstanceOf[Byte])
+
+      case (INT, ShortType) =>
+        (updater, ordinal, value) => updater.setShort(ordinal, value.asInstanceOf[Short])
+
+      case (INT, ArrayType(IntegerType, containsNull)) =>
```

Comment:
Is there a way to detect if the proto type is repeated?

Reply:
@gengliangwang yes, The protobuf descriptor is used to build the catalyst schema. However, we can still make modifications like the one below. Do you believe it to be an improvement over the prior strategy?

```
      case  (INT, ArrayType(IntegerType, containsNull)) if protoType.isRepeated =>
        newArrayWriter(protoType, protoPath, catalystPath, IntegerType, containsNull)
```

Reply:
I think we can avoid duplicated code for the array branches. It seems that we can avoid so many `newArrayWriter` method calls.

Reply:
Again, feel free to improve it in a followup PR

Reply:
@gengliangwang made changes.